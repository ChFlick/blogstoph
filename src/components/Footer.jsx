import React from 'react';

const Footer = (props) => (
    <footer>
        <hr className="bar--light-grey" />
        <div className="copyright">Copyright Â© Christoph Flick 2018</div>
    </footer>
);

export default Footer;
