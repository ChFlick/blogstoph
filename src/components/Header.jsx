import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';

import { startLogout } from '../actions/auth';

const DEFAULT_HEADER_CLASSES = 'header header--fixed-top';

export class Header extends React.Component {
    state = {
        headerClasses: DEFAULT_HEADER_CLASSES
    }

    componentDidMount = () => {
        window.addEventListener('scroll', this.handleScroll);
    };

    componentWillUnmount = () => {
        window.removeEventListener('scroll', this.handleScroll);
    };

    handleScroll = () => {
        const scrollPosPrev = this.state.scrollPos;
        const scrollPos = window.scrollY;
        let headerClasses = 'header header--fixed-top';

        if (scrollPos > 80) {
            headerClasses = 'header header--fixed';
            if (scrollPos < scrollPosPrev) {
                headerClasses += ' header--visible';
            }

            this.setState(() => ({ headerClasses }));
        } else if (scrollPos === 0) {
            this.setState(() => ({ headerClasses: DEFAULT_HEADER_CLASSES }));
        }

        this.setState(() => ({ scrollPos }));
    };

    createAreaSwitchButton = () =>
        this.props.private ?
            <Link className="button button--link button--header" to="/dashboard">Visitor Area</Link> :
            <Link className="button button--link button--header" to="/editor/dashboard">Editor Area</Link>;


    render() {
        return (
            <header id="header" className={this.state.headerClasses} onScroll={this.handleScroll}>
                <div className="content-container">
                    <div className="header__content">
                        <Link className="header__title" to="/dashboard">
                            <h1>Blogstoph</h1>
                        </Link>
                        {this.props.isAuthenticated ? (
                            <div>
                                {this.createAreaSwitchButton()}
                                <button className="button button--link button--header" onClick={this.props.startLogout}>Logout</button>
                            </div>
                        ) : (
                                <Link className="button button--link" to="/login">Login</Link>
                            )}
                    </div>
                </div>
            </header>
        );
    };
};

const mapStateToProps = (state) => ({
    isAuthenticated: !!state.auth.uid
});

const mapDispatchToProps = (dispatch) => ({
    startLogout: () => dispatch(startLogout()),
});

export default connect(mapStateToProps, mapDispatchToProps)(Header);
