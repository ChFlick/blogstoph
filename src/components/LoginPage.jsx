import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';

export const LoginPage = ({ startLogin }) => (
    <div className="box-layout">
        <div className="box-layout__box">
            <h1 className="box-layout__title">Blogstoph Login</h1>
            <p>Log in to edit your posts.</p>
            <button className="button" onClick={startLogin}>Login with Google</button>
        </div>
    </div>
);

const mapDispatchToProps = (dispatch) => ({
    startLogin: () => dispatch(startLogin('google'))
});

export default connect(undefined, mapDispatchToProps)(LoginPage);
