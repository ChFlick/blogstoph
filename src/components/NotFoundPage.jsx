import React, { Fragment } from 'react';
import { Link } from 'react-router-dom';

const NotFoundPage = () => (
    <Fragment>
        <header className="page-header">
            <h1>404 Not Found</h1>
        </header>
        <section className="content-container align-center">
            <p>The page you were looking for could not be found.</p>
            <Link to="/">Click here to go back to the home page</Link>
        </section>
    </Fragment>
);

export default NotFoundPage;
