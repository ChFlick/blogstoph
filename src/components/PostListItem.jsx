import React, { Fragment } from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';

const PostListItem = (props) => (
    <Fragment>
        <Link
            className="post-preview"
            to={`post/${props.post.id}`}
        >
            <h3 className="post-preview__title">{props.post.title}</h3>
            <p className="post-preview__subtitle">{props.post.content}</p>
        </Link>
        <span className="post-preview__data">
            Posted by {props.post.author} at {moment(props.post.date).format('DD. MMM YYYY')}
        </span>
    </Fragment>
);

export default PostListItem;
