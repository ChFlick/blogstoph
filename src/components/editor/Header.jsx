import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';

import { startLogout } from '../../actions/auth';


export const EditorHeader = ({ startLogout }) => (
    <header className="header">
        <div className="content-container">
            <div className="header__content">
                <Link className="header__title" to="/editor/dashboard">
                    <h1>Blogstoph</h1>
                </Link>
                <button className="button button--link" onClick={startLogout}>Logout</button>
            </div>
        </div>
    </header>
);

const mapDispatchToProps = (dispatch) => ({
    startLogout: () => dispatch(startLogout())
});

export default connect(undefined, mapDispatchToProps)(EditorHeader);
