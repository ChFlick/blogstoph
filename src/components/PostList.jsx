import React, { Fragment } from 'react';
import { connect } from 'react-redux';

import PostListItem from './PostListItem';

const isNotLastElement = (index, numberOfPosts) => {
    return index !== numberOfPosts - 1;
};

export const PostList = (props) => (
    <div className="content-container">
        {
            props.posts.length === 0 ? (
                <p>There are no posts available</p>
            ) : (
                props.posts.map((post, index) =>
                    <Fragment key={post.id}>
                        <PostListItem post={post} />
                        {isNotLastElement(index, props.posts.length) && <hr className="bar--light-grey" />}
                    </Fragment>
                )
            )
        }
    </div>
);

const mapStateToProps = (state) => ({
    posts: (state.posts || []).filter((post = {}) => post.published === true)
});

export default connect(mapStateToProps)(PostList);
