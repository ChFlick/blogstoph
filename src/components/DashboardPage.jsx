import React from 'react';

import PostList from './PostList';

const DashboardPage = () => (
    <section>
        <header className="page-header">
            <div className="content-container">
                <h1 className="page-header__title page-header__title--xl">Blogstoph</h1>
                <hr className="page-header__bar"/>
                <h2 className="page-header__subtitle">Christoph Flicks IT blog</h2>
            </div>
        </header>
        <PostList />
    </section>
);

export default DashboardPage;
