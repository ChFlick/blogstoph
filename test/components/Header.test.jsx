import React from 'react';
import { shallow, mount } from 'enzyme';

import { Header } from '../../src/components/Header';


test('should render Header correctly when logged out', () => {
    const wrapper = shallow(<Header startLogout={() => {}} isAuthenticated={false} />);

    expect(wrapper).toMatchSnapshot();
});

test('should render Header correctly when logged in', () => {
    const wrapper = shallow(<Header startLogout={() => {}} isAuthenticated={true} />);

    expect(wrapper).toMatchSnapshot();
});


test('should render Header correctly when logged in in private area', () => {
    const wrapper = shallow(<Header private startLogout={() => {}} isAuthenticated={true} />);

    expect(wrapper).toMatchSnapshot();
});


test('should call startLogout on button click', () => {
    const startLogout = jest.fn();
    const wrapper = shallow(<Header startLogout={startLogout} isAuthenticated={true} />);

    wrapper.find('button').simulate('click');

    expect(startLogout).toHaveBeenCalled();
});

test('should set scrollPos in onScroll', () => {
    const scrollPosVal = 33;
    global.scrollY = scrollPosVal;

    const wrapper = shallow(<Header />);

    wrapper.prop('onScroll')();

    expect(wrapper.state('scrollPos')).toEqual(scrollPosVal);
});

test('should set header fixed if scrolled down', () => {
    global.scrollY = 500;

    const wrapper = shallow(<Header />);

    wrapper.prop('onScroll')();

    expect(wrapper.state('headerClasses')).toEqual('header header--fixed');
});

test('should set header fixed and visible if scrolled down and up again', () => {
    const scrollPosVal = 500;
    global.scrollY = scrollPosVal;

    const wrapper = shallow(<Header />);
    wrapper.setState({scrollPos: scrollPosVal + 100 });
    wrapper.prop('onScroll')();

    expect(wrapper.state('headerClasses')).toEqual('header header--fixed header--visible');
});

test('should set header back to top if scoll 0', () => {
    const scrollPosVal = 0;
    global.scrollY = scrollPosVal;

    const wrapper = shallow(<Header />);
    wrapper.prop('onScroll')();

    expect(wrapper.state('headerClasses')).toEqual('header header--fixed-top');
});

test('can unmound', () => {
    const wrapper = shallow(<Header />);
    wrapper.unmount();
});
