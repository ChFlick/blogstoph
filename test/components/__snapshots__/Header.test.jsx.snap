// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Header correctly when logged in 1`] = `
<header
  className="header header--fixed-top"
  id="header"
  onScroll={[Function]}
>
  <div
    className="content-container"
  >
    <div
      className="header__content"
    >
      <Link
        className="header__title"
        replace={false}
        to="/dashboard"
      >
        <h1>
          Blogstoph
        </h1>
      </Link>
      <nav>
        <Link
          className="button button--link button--header"
          replace={false}
          to="/editor/dashboard"
        >
          Editor Area
        </Link>
        <button
          className="button button--link button--header"
          onClick={[Function]}
        >
          Logout
        </button>
      </nav>
    </div>
  </div>
</header>
`;

exports[`should render Header correctly when logged in in private area 1`] = `
<header
  className="header header--fixed-top"
  id="header"
  onScroll={[Function]}
>
  <div
    className="content-container"
  >
    <div
      className="header__content"
    >
      <Link
        className="header__title"
        replace={false}
        to="/dashboard"
      >
        <h1>
          Blogstoph
        </h1>
      </Link>
      <nav>
        <Link
          className="button button--link button--header"
          replace={false}
          to="/dashboard"
        >
          Visitor Area
        </Link>
        <button
          className="button button--link button--header"
          onClick={[Function]}
        >
          Logout
        </button>
      </nav>
    </div>
  </div>
</header>
`;

exports[`should render Header correctly when logged out 1`] = `
<header
  className="header header--fixed-top"
  id="header"
  onScroll={[Function]}
>
  <div
    className="content-container"
  >
    <div
      className="header__content"
    >
      <Link
        className="header__title"
        replace={false}
        to="/dashboard"
      >
        <h1>
          Blogstoph
        </h1>
      </Link>
      <Link
        className="button button--link"
        replace={false}
        to="/login"
      >
        Login
      </Link>
    </div>
  </div>
</header>
`;
